File: Backend/app/models/season.py
from __future__ import annotations

from typing import TYPE_CHECKING
from datetime import date

from sqlalchemy import ForeignKeyConstraint, PrimaryKeyConstraint, String, Text, Date, Float
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.db import Base

if TYPE_CHECKING:
    from app.models import Media, Episode


class Season(Base):
    __tablename__ = "seasons"

    media_id: Mapped[int] = mapped_column(nullable=False)
    season_number: Mapped[int] = mapped_column(nullable=False)

    title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    overview: Mapped[str | None] = mapped_column(Text, nullable=True)
    release_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    tmdb_rating: Mapped[float | None] = mapped_column(Float, nullable=True)

    __table_args__ = (
        PrimaryKeyConstraint("media_id", "season_number"),
        ForeignKeyConstraint(
            ["media_id"],
            ["media.id"],
            ondelete="CASCADE",
        ),
    )

    media: Mapped["Media"] = relationship(back_populates="seasons")
    episodes: Mapped[list["Episode"]] = relationship(
        back_populates="season",
        cascade="all, delete-orphan",
    )

File: movie_tracker_app/lib/models/series/season.dart
import 'episode.dart';

class Season {
  final int seasonNumber;
  final String? title;
  final String? overview;
  final DateTime? releaseDate;
  final double? tmdbRating;
  final List<Episode> episodes;

  const Season({
    required this.seasonNumber,
    this.title,
    this.overview,
    this.releaseDate,
    this.tmdbRating,
    this.episodes = const [],
  });

  factory Season.fromJson(Map<String, dynamic> json) {
    return Season(
      seasonNumber: json['season_number'],
      title: json['title'],
      overview: json['overview'],
      releaseDate: json['release_date'] != null
          ? DateTime.parse(json['release_date'])
          : null,
      tmdbRating: (json['tmdb_rating'] as num?)?.toDouble(),
      episodes: (json['episodes'] as List<dynamic>?)
              ?.map((e) => Episode.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
    'season_number': seasonNumber,
    'title': title,
    'overview': overview,
    'release_date': releaseDate?.toIso8601String(),
    'tmdb_rating': tmdbRating,
    'episodes': episodes.map((e) => e.toJson()).toList(),
  };

  Season copyWith({
    int? seasonNumber,
    String? title,
    String? overview,
    DateTime? releaseDate,
    double? tmdbRating,
    List<Episode>? episodes,
  }) {
    return Season(
      seasonNumber: seasonNumber ?? this.seasonNumber,
      title: title ?? this.title,
      overview: overview ?? this.overview,
      releaseDate: releaseDate ?? this.releaseDate,
      tmdbRating: tmdbRating ?? this.tmdbRating,
      episodes: episodes ?? this.episodes,
    );
  }
}